nova_vm/src/ecmascript/types/language/number.rs

```diff
@@ -76,13 +103,8 @@ impl TryFrom<Value> for Number {
 }
 
 impl Number {
-    pub fn new(value: Value) -> Self {
-        debug_assert!(matches!(
-            value,
-            Value::Number(_) | Value::Integer(_) | Value::Float(_)
-        ));
-        // SAFETY: Sub-enum.
-        unsafe { std::mem::transmute::<Value, Number>(value) }
```

Comment:
As said, TryFrom cannot get an agent or heap reference so that won't work. You're definitely right though that the order of things here is really odd.

Reply:
I ended up duplicating the code from `CreateHeapData` macro here. The reason I didn't just move the code here is that the `CreateHeapData` function doesn't have access to `Agent` which we currently take in all our public APIs.